File: backend/api/schemas/onboarding.py
"""
Onboarding schemas.
Handles user onboarding flow including style quiz and preference setup.
"""

from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field, field_validator


class OnboardingProductsRequest(BaseModel):
    """Request for fetching products for onboarding moodboard."""
    limit: int = Field(20, ge=12, le=50, description="Number of products to return")
    diverse: bool = Field(True, description="Whether to return diverse products across categories")


class OnboardingProduct(BaseModel):
    """Simplified product for onboarding display."""
    product_id: str
    title: str
    image_url: Optional[str]
    price: float
    brand: Optional[str]
    category: Optional[str]


class OnboardingProductsResponse(BaseModel):
    """Response containing products for moodboard selection."""
    products: List[OnboardingProduct]
    total: int


class OnboardingCompleteRequest(BaseModel):
    """Request to complete onboarding with style preferences."""
    selected_product_ids: List[str] = Field(
        ...,
        min_length=3,
        max_length=5,
        description="3-5 product IDs selected from moodboard"
    )
    price_min: Optional[float] = Field(None, ge=0, description="Minimum price preference")
    price_max: Optional[float] = Field(None, ge=0, description="Maximum price preference")

    @field_validator('price_max')
    @classmethod
    def validate_price_range(cls, v, info):
        """Ensure max price is greater than min price."""
        if v is not None and info.data.get('price_min') is not None:
            if v <= info.data['price_min']:
                raise ValueError("Maximum price must be greater than minimum price")
        return v


class OnboardingCompleteResponse(BaseModel):
    """Response after completing onboarding."""
    success: bool
    user_id: str
    onboarded: bool
    embedding_created: bool
    preferences_saved: bool
    selected_products_count: int
    message: str
    next_step: str = Field("/", description="Where to redirect user after onboarding")

    # Metadata about the created embedding
    embedding_metadata: Optional[Dict[str, Any]] = Field(
        None,
        description="Information about the created embedding (confidence, method, etc.)"
    )


class OnboardingStatusResponse(BaseModel):
    """Response for checking onboarding status."""
    user_id: str
    onboarded: bool
    has_embedding: bool
    has_preferences: bool
    registration_date: str
    onboarding_date: Optional[str] = None

File: frontend/src/lib/queries/onboarding.ts
/**
 * Onboarding Query Hooks
 */

import { useQuery, useMutation, UseQueryOptions, UseMutationOptions } from "@tanstack/react-query";

const API_URL = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8001";

/**
 * Product for onboarding moodboard
 */
export interface OnboardingProduct {
  product_id: string;
  title: string;
  image_url: string | null;
  price: number;
  brand: string | null;
  category: string | null;
}

/**
 * Response from onboarding products endpoint
 */
export interface OnboardingProductsResponse {
  products: OnboardingProduct[];
  total: number;
}

/**
 * Request to complete onboarding
 */
export interface OnboardingCompleteRequest {
  selected_product_ids: string[];
  price_min: number | null;
  price_max: number | null;
}

/**
 * Response from complete onboarding endpoint
 */
export interface OnboardingCompleteResponse {
  success: boolean;
  user_id: string;
  onboarded: boolean;
  embedding_created: boolean;
  preferences_saved: boolean;
  selected_products_count: number;
  message: string;
  next_step: string;
  embedding_metadata?: {
    confidence: number;
    method: string;
    products_used: number;
  };
}

/**
 * Hook to get products for onboarding moodboard
 */
export function useOnboardingProducts(
  options?: Omit<UseQueryOptions<OnboardingProductsResponse>, "queryKey" | "queryFn">
) {
  return useQuery({
    queryKey: ["onboarding", "products"],
    queryFn: async (): Promise<OnboardingProductsResponse> => {
      const response = await fetch(
        `${API_URL}/api/v1/onboarding/products?limit=20&diverse=true`,
        {
          method: "GET",
          credentials: "include",
        }
      );

      if (!response.ok) {
        const error = await response.json();
        throw new Error(error.message || "Failed to load onboarding products");
      }

      return response.json();
    },
    staleTime: 1000 * 60 * 10, // 10 minutes
    ...options,
  });
}

/**
 * Hook to complete onboarding with style preferences
 */
export function useCompleteOnboarding(
  options?: Omit<
    UseMutationOptions<OnboardingCompleteResponse, Error, OnboardingCompleteRequest>,
    "mutationFn"
  >
) {
  return useMutation({
    mutationFn: async (
      request: OnboardingCompleteRequest
    ): Promise<OnboardingCompleteResponse> => {
      const response = await fetch(`${API_URL}/api/v1/onboarding/complete`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(request),
        credentials: "include",
      });

      if (!response.ok) {
        const error = await response.json();
        throw new Error(error.detail || error.message || "Failed to complete onboarding");
      }

      return response.json();
    },
    ...options,
  });
}
